Implementation of intra-broker goal violation detection and self-healing

dgldn opened the pull request:
## Summary
This feature is completely based on work from @ecojan who worked on the [PR](https://github.com/linkedin/cruise-control/pull/1721). And, reviewed previoulsy by @alexander-falca @amuraru @zornhsu @CCisGG

1. Why:
This PR resolves [#1264](https://github.com/linkedin/cruise-control/issues/1264) by the goal violation detection and self-healing for intra-broker goals. This however does not account for race conditions between Cluster level Goals and will only trigger after said goals are finished.
The scope of this PR is to allow cruise control run IntraBroker goal violation detection and execute self-healing for now (as future work might make this obsolete as intra-broker goals and regular goals might be made to work together).

2. What:
Implementation of intra-broker goals violation detection and executing self-healing in case of JBOD(multiple logDirs) based kafka clusters.

### New Configurations:
```
anomaly.detection.intra.broker.goals= List of intra-broker goals for the Anomaly detector to look for
self.healing.intra.broker.goals= List of Intra broker Goals for self-healing
self.healing.intra.broker.goal.violation.enabled= If the self-healing should be enabled for intra-broker-goals. This configuration defaults to self.healing.enabled
```

The configs `anomaly.detection.intra.broker.goals` and `self.healing.intra.broker.goals` both default to `com.linkedin.kafka.cruisecontrol.analyzer.goals.IntraBrokerDiskCapacityGoal,com.linkedin.kafka.cruisecontrol.analyzer.goals.IntraBrokerDiskUsageDistributionGoal` since these are the only cruise control intra-broker goals available at the moment.

The `self.healing.intra.broker.goal.violation.enabled` can be disbaled explicilty overriding the config to `false`.

### New Anomaly Type: INTRA_BROKER_GOAL_VIOLATION
A new type of Anomaly that needs to be treated differently from the regular GOAL_VIOLATIONS.
When configuring for the rebalance parameters for this violation, we are skipping hard goals check and ignoring proposalsCache.

We are skipping hard goals check because intra-broker goals may not be hard goals (at the current given time).
We are ignoring proposalsCache because the proposals model takes into account all given default goals. If we add the intraBrokersGoals this will not work properly as currently, disk-granularity goals do not work with broker-granularity goals.

### New Anomaly detectors: IntraBrokerGoalViolationDetector
This anomaly detector is only looking for the anomaly.detection.intra.broker.goal and unlike the GoalViolationDetector, it's using a cluster model that is generating replica placement on disks as well.
This anomaly detector mainly functions for INTRA_BROKER_GOAL_VIOLATIONS.

### What is missing from this feature
Missing representation on INTRA_BROKER_GOAL_VIOLATIONS in CruiseControl UI when an anomaly is detected (as this is a structure in cruise-control-ui project)

## Expected Behavior
In case of JBOD kafka clusters, IntraBroker goal violation detection and self-healing should be excuted by cruise control  similar to Inter-Broker goal violations without any user inputs.

## Known Workarounds
Workaround is the existing feature of manually executing `rebalance` API with `rebalance_disk=true` or using the cruise control UI

## Categorization
- [x] documentation
- [ ] bugfix
- [x] new feature
- [ ] refactor
- [ ] security/CVE
- [ ] other

This PR resolves #1264.

<img width="1512" height="835" alt="image" src="https://github.com/user-attachments/assets/64c5da7a-0494-4451-b301-d2cda43ebc4f" />

Logs of cruise control execution against a 3 broker kafka cluster: [kafkacruisecontrol.log](https://github.com/user-attachments/files/21300692/kafkacruisecontrol.log)

dattagiri committed:
Implementation of intra-broker goal violation detection and self-healing

dattagiri committed:
Updating README

ecojan commented:
Thank you for picking this up @dgldn!

kyguy reviewed (commented):
Hi @dgldn, just had an initial pass and left a couple comments. It appears the maintainers are a bit swamped so the bulk of the code review is probably going to have to come from volunteers in the community. To save time on reviews, it would be great to double check that this PR addresses all the feedback left on the original! [1] 

[1] https://github.com/linkedin/cruise-control/pull/1721

egyedt committed:
Remove ZooKeeper (#2244)

* Remove ZooKeeper from Cruise Control production

Authored-by: Tamas Barnabas Egyed <egyed.t@cloudera.com>
Change-Id: Ic9a7e255193cfbbc6c537e5138c86725363a74b9

* Use KRaft in Cruise Control tests

Co-authored-by: Patrik Marton <pmarton@cloudera.com>
Co-authored-by: Tamas Barnabas Egyed <egyed.t@cloudera.com>

---------

Co-authored-by: Patrik Marton <pmarton@cloudera.com>
# Conflicts:
#	cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManager.java
#	docs/wiki/User Guide/Configurations.md

# Conflicts:
#	README.md
#	build.gradle
#	cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporterTest.java
#	cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedKRaftController.java
#	cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCKafkaIntegrationTestHarness.java
#	cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCKafkaRaftServer.java
#	cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetectorManager.java
#	cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUnitTestUtils.java
#	cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/GoalOptimizerTest.java
#	cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/config/BrokerSetFileResolverTest.java
#	cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/config/CaseInsensitiveGoalConfigTest.java
#	cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ConcurrencyAdjusterTest.java
#	gradle.properties

dgldn commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractGoal.java:
```diff
@@ -433,4 +433,34 @@ Replica maybeSwapReplicaBetweenDisks(ClusterModel clusterModel,
+
+  // Overriding equals() to compare two AbstractGoals objects
+  @Override
+  public boolean equals(Object o) {
```
This method as well is not used anywhere.

dgldn commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerCapacityConfigFileResolver.java:
```diff
@@ -188,6 +188,17 @@ public BrokerCapacityInfo capacityForBroker(String rack, String host, int broker
+      }
+    }
+    return true;
+  }
```
@kyguy , will move the logic to partial JBOD cluster too. Since there are sceanrios where only a few brokers have multiple logdirs set up. Thank you for pointing it out.

dattagiri committed:
Changing the JBOS evaluation to partial JBOD kafka cluster too

dgldn committed:
Merge branch 'main' into intra-broker-self-healing-disk-rebalance

dgldn committed:
Update Configurations.md to add intra broker self healing related configs

dattagiri committed:
Update Configurations.md to add intra broker self healing related configs

egyedt commented:
This seems to be a good feature and it would be nice to have it finally in Cruise Control!
@kyguy Do you have any additional comments, or can you approve this PR? You already reviewed it.
@bgrishinko Could you please take a look and speed up the process somehow if possible?
Thanks in advance!

kyguy commented:
> Do you have any additional comments, or can you approve this PR? You already reviewed it.

If I remember correctly there were some outstanding comments from the original PR [1] that still need to be addressed. I'll have another pass this week to confirm!

[1] https://github.com/linkedin/cruise-control/pull/1721

dgldn committed:
Merge branch 'linkedin:main' into intra-broker-self-healing-disk-rebalance

dgldn committed:
Refactoring the GoalViolationDetector to include both inter-broker and intra-broker goal violation detectors

dgldn commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/IntraBrokerGoalViolationDetector.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    @Override
+    public void run() {
```
This has been addressed by combining GoalViolationDetector and IntraBrokerGoalViolationDetector

dgldn commented:
The PR has been updated to combine both inter-broker and intra-broker goal violation detectors into a single class `GoalViolationDetector.java` which addresses the duplicated methods used separately in `GoalViolationDetector.java` and `IntraBrokerGoalViolationDetector.java`
Removed the `BaseGoalViolationDetector.java` and `IntraBrokerGoalViolationDetector.java` after refactoring.
